In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
%matplotlib inline

In [2]:
df = pd.read_csv('House-Price (1).csv')
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [3]:
x = df.drop(['id','date','price'],axis=1)
y = df['price']

In [5]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.2,random_state = 100) 

In [6]:
from sklearn.linear_model import LinearRegression
lm = LinearRegression()
lm = lm.fit(x_train,y_train)

In [7]:
coef = pd.concat([pd.DataFrame(x_train.columns),pd.DataFrame(np.transpose(lm.coef_))], axis = 1)

In [8]:
coef

0              0
0        bedrooms  -35079.863357
1       bathrooms   47484.794554
2     sqft_living     106.601448
3        sqft_lot       0.104526
4          floors    6795.132057
5      waterfront  590318.794222
6            view   52019.494721
7       condition   27573.775394
8           grade   97450.025997
9      sqft_above      70.821265
10  sqft_basement      35.780182
11       yr_built   -2750.360996
12   yr_renovated      10.061625
13        zipcode    -555.872473
14            lat  600380.157582
15           long -206135.142537
16  sqft_living15      21.174889
17     sqft_lot15      -0.344935

In [9]:
y_pred = lm.predict(x_test)

In [11]:
y_error = y_test - y_pred
y_error

19836    -88019.801836
10442    -48655.657207
20548    -88877.852737
11014   -103654.040623
4138      60085.433313
             ...      
5625      -8674.866417
8698     -75359.290473
9786     159093.733618
2784    -148122.857648
9552      68585.882305
Name: price, Length: 4323, dtype: float64

In [12]:
from sklearn.metrics import r2_score
r2_score(y_test,y_pred)

0.7028918252460872

In [13]:
import statsmodels.api as sma
X_train = sma.add_constant(x_train) ## let's add an intercept (beta_0) to our model
X_test = sma.add_constant(x_test)

C:\Users\Ashish\Anaconda3\envs\python\lib\site-packages\numpy\core\fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [19]:
lm2 = sma.OLS(y_train,X_train).fit()

In [20]:
lm2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.699
Model:                            OLS   Adj. R-squared:                  0.698
Method:                 Least Squares   F-statistic:                     2356.
Date:                Wed, 01 Jul 2020   Prob (F-statistic):               0.00
Time:                        13:58:24   Log-Likelihood:            -2.3572e+05
No. Observations:               17290   AIC:                         4.715e+05
Df Residuals:                   17272   BIC:                         4.716e+05
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          5.485e+06   3.31e+06      1.659      0.097   -9.97e+05     1.2e+07
bedrooms      -3.508e+04   2102.139    -16.688      0.000   -3.92e+04    -3.1e+04
bathrooms      4.748e+04   3660.769     12.971      0.000    4.03e+04    5.47e+04
sqft_living     106.6002      2.566     41.536      0.000     101.570     111.631
sqft_lot          0.1045      0.051      2.061      0.039       0.005       0.204
floors         6795.1321   4037.885      1.683      0.092   -1119.531    1.47e+04
waterfront     5.903e+05    1.9e+04     30.997      0.000    5.53e+05    6.28e+05
view           5.202e+04   2411.649     21.570      0.000    4.73e+04    5.67e+04
condition      2.757e+04   2637.356     10.455      0.000    2.24e+04    3.27e+04
grade          9.745e+04   2399.368     40.615      0.000    9.27e+04    1.02e+05
sqft_above       70.8225      2.547     27.811      0.000      65.831      75.814
sqft_basement    35.7814      2.985     11.989      0.000      29.931      41.631
yr_built      -2750.3610     81.011    -33.950      0.000   -2909.152   -2591.570
yr_renovated     10.0616      4.133      2.435      0.015       1.961      18.162
zipcode        -555.8725     37.214    -14.937      0.000    -628.816    -482.929
lat            6.004e+05    1.2e+04     49.881      0.000    5.77e+05    6.24e+05
long          -2.061e+05   1.47e+04    -14.070      0.000   -2.35e+05   -1.77e+05
sqft_living15    21.1749      3.885      5.451      0.000      13.561      28.789
sqft_lot15       -0.3449      0.078     -4.417      0.000      -0.498      -0.192
==============================================================================
Omnibus:                    14869.428   Durbin-Watson:                   2.036
Prob(Omnibus):                  0.000   Jarque-Bera (JB):          1526532.644
Skew:                           3.627   Prob(JB):                         0.00
Kurtosis:                      48.457   Cond. No.                     1.30e+17
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.03e-20. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [21]:
y_pred2 = lm2.predict(X_test) 

Note that both y_pred and y_pred2 are same. It's just these are calculated via different packages.

In [23]:
x =y_pred - y_pred2
x

19836   -6.752089e-09
10442    2.335291e-07
20548    8.732313e-07
11014    9.546056e-08
4138     1.674052e-07
             ...     
5625     1.595763e-07
8698     5.844049e-08
9786     8.940697e-08
2784    -2.363231e-07
9552    -8.614734e-08
Length: 4323, dtype: float64

In [26]:
y_test = pd.to_numeric(y_test, errors='coerce')
RSS = np.sum((y_pred2 - y_test)**2)
y_mean = np.mean(y_test)
TSS = np.sum((y_test - y_mean)**2)
R2 = 1 - RSS/TSS
print(R2)

n=X_test.shape[0]
p=X_test.shape[1] - 1

adj_rsquared = 1 - (1 - R2) * ((n - 1)/(n-p-1))
print(adj_rsquared)

0.7028918252460696
0.7016492724706117


## Detecting Outliers:
Firstly we try to get the studentized residuals using get_influence( ). The studentized residuals are saved in resid_student.

In [32]:
influence = lm2.get_influence()  
influence

In [34]:
from statsmodels.stats.outliers_influence import OLSInfluence

In [39]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
[variance_inflation_factor(x_train.values, j) for j in range(x_train.shape[1])]

C:\Users\Ashish\Anaconda3\envs\python\lib\site-packages\statsmodels\stats\outliers_influence.py:185: RuntimeWarning: divide by zero encountered in double_scalars
  vif = 1. / (1. - r_squared_i)


[22.819267327799544,
 28.725038659156816,
 inf,
 2.2511936351457633,
 16.98665877618419,
 1.2297035823145008,
 1.5741318083007199,
 34.823471185867916,
 145.74224575686694,
 inf,
 inf,
 9473.271956354276,
 1.1933835378689206,
 1617923.4026960507,
 139088.1121297648,
 1347819.5729790383,
 27.36487713715211,
 2.4581244798451314]

We create a function to remove the collinear variables. We choose a threshold of 5 which means if VIF is more than 5 for a particular variable then that variable will be removed.

In [40]:
def calculate_vif(x):
    thresh = 5.0
    output = pd.DataFrame()
    k = x.shape[1]
    vif = [variance_inflation_factor(x.values, j) for j in range(x.shape[1])]
    for i in range(1,k):
        print("Iteration no.")
        print(i)
        print(vif)
        a = np.argmax(vif)
        print("Max VIF is for variable no.:")
        print(a)
        if vif[a] <= thresh :
            break
        if i == 1 :          
            output = x.drop(x.columns[a], axis = 1)
            vif = [variance_inflation_factor(output.values, j) for j in range(output.shape[1])]
        elif i > 1 :
            output = output.drop(output.columns[a],axis = 1)
            vif = [variance_inflation_factor(output.values, j) for j in range(output.shape[1])]
    return(output)
train_out = calculate_vif(x_train) 

Iteration no.
1
[22.819267327799544, 28.725038659156816, inf, 2.2511936351457633, 16.98665877618419, 1.2297035823145008, 1.5741318083007199, 34.823471185867916, 145.74224575686694, inf, inf, 9473.271956354276, 1.1933835378689206, 1617923.4026960507, 139088.1121297648, 1347819.5729790383, 27.36487713715211, 2.4581244798451314]
Max VIF is for variable no.:
2
Iteration no.
2
[22.819267327799544, 28.725038659156816, 2.2511936351457633, 16.98665877618422, 1.2297035823145004, 1.5741318083007205, 34.823471185867916, 145.74224575686694, 28.20363073387459, 2.8646396931983147, 9473.271956354276, 1.1933835378689202, 1617923.4026960507, 139088.1121297648, 1347819.5729790383, 27.36487713715211, 2.4581244798451296]
Max VIF is for variable no.:
12
Iteration no.
3
[22.819246745608886, 28.65066407535827, 2.239105337200614, 16.830529909358035, 1.2294203472922816, 1.5740577908695441, 34.78473570082762, 139.98232972137416, 27.722931768294625, 2.8515401767929998, 8434.29410896354, 1.1882782516356238, 11973

In [41]:
train_out.head()

sqft_lot  waterfront  view  sqft_above  sqft_basement  yr_renovated  \
16000     11096           0     0        1210            570             0   
11286      2229           0     0        1420              0             0   
3201       4000           0     0        1040            750             0   
11049    117176           0     0        1710              0             0   
9716       5217           0     0        1150              0             0   

       sqft_lot15  
16000       10640  
11286        2230  
3201         4000  
11049       12223  
9716         5217

In [45]:
x_test.head()
x_test.drop(['bedrooms','bathrooms','sqft_living','floors','condition','grade','yr_built','zipcode','lat','long','sqft_living15'],axis = 1,inplace = True)

KeyError: "['bedrooms' 'bathrooms' 'sqft_living' 'floors' 'condition' 'grade'\n 'yr_built' 'zipcode' 'lat' 'long' 'sqft_living15'] not found in axis"

In [46]:
x_test.head()

sqft_lot  waterfront  view  sqft_above  sqft_basement  yr_renovated  \
19836      5136           0     0        2437              0             0   
10442      4800           0     0        1560              0             0   
20548     47480           0     0        2390              0             0   
11014      5100           0     0        1480              0          1959   
4138       1341           0     0        1650            230             0   

       sqft_lot15  
19836        4614  
10442       12240  
20548       44866  
11014        5100  
4138         1883

In [48]:
import statsmodels.api as sma
import statsmodels.formula.api as sm
train_out = sma.add_constant(train_out) ## let's add an intercept (beta_0) to our model
X_test = sma.add_constant(x_test)
lm2 = sma.OLS(y_train,train_out).fit()
lm2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.553
Model:                            OLS   Adj. R-squared:                  0.552
Method:                 Least Squares   F-statistic:                     3049.
Date:                Wed, 01 Jul 2020   Prob (F-statistic):               0.00
Time:                        14:48:24   Log-Likelihood:            -2.3914e+05
No. Observations:               17290   AIC:                         4.783e+05
Df Residuals:                   17282   BIC:                         4.783e+05
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const         -1.823e+04   4736.935     -3.849      0.000   -2.75e+04   -8948.558
sqft_lot         -0.0115      0.061     -0.187      0.852      -0.132       0.109
waterfront     5.657e+05   2.32e+04     24.427      0.000     5.2e+05    6.11e+05
view           7.608e+04   2830.523     26.878      0.000    7.05e+04    8.16e+04
sqft_above      263.1118      2.361    111.455      0.000     258.485     267.739
sqft_basement   245.6086      4.447     55.230      0.000     236.892     254.325
yr_renovated     49.0902      4.736     10.366      0.000      39.808      58.372
sqft_lot15       -0.7038      0.094     -7.483      0.000      -0.888      -0.519
==============================================================================
Omnibus:                    10895.228   Durbin-Watson:                   2.026
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           397160.383
Skew:                           2.481   Prob(JB):                         0.00
Kurtosis:                      25.950   Cond. No.                     6.43e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.43e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Checking normality of residuals We use Shapiro Wilk test from scipy library to check the normality of residuals.
Null Hypothesis: The residuals are normally distributed.
Alternative Hypothesis: The residuals are not normally distributed.

In [49]:
from scipy import stats
stats.shapiro(lm2.resid)

C:\Users\Ashish\Anaconda3\envs\python\lib\site-packages\scipy\stats\morestats.py:1676: UserWarning: p-value may not be accurate for N > 5000.
  warnings.warn("p-value may not be accurate for N > 5000.")


(0.8586614727973938, 0.0)

Since the pvalue is 0.0 thus at 5% level of significance we can say that the residuals are not normally distributed.

## Checking for autocorrelation To ensure the absence of autocorrelation we use Ljungbox test.
Null Hypothesis: Autocorrelation is absent.
Alternative Hypothesis: Autocorrelation is present.

In [51]:
from statsmodels.stats import diagnostic as diag
diag.acorr_ljungbox(lm2.resid , lags = 1) 

(array([2.87681097]), array([0.08986383]))

Since p-value is 0.089 thus we can accept the null hypothesis and can say that autocorrelation is absent.

## `Checking heteroscedasticity Using Goldfeld Quandt we test for heteroscedasticity.
Null Hypothesis: Error terms are homoscedastic
Alternative Hypothesis: Error terms are heteroscedastic.

In [54]:
import statsmodels.stats.api as sms
from statsmodels.compat import lzip
name = ['F statistic', 'p-value']
test = sms.het_goldfeldquandt(lm2.resid, lm2.model.exog)
lzip(name, test)

[('F statistic', 0.946846717280032), ('p-value', 0.9944215562569185)]

The p-value is 0.994 hence we can say that the residuals have constant variance. Hence we can say that all the assumptions of our linear regression model are satisfied.